### ISSUE TITLE
Decrypt LE Secure Connection with known LTK does not work

### ISSUE THREAD

**Issue by jkrauss1987** [mikeryan/crackle#53]
Hi, I have a pcap file from sniffing an LE Secure owner pairing. I have got the correct LTK on another way. 
The output of the tool reports a successful decription and a new output.pcap file is generated. But when i load it into Wireshark the payload of the decypted packets is much less than in the original files messages and Wireshark says "bad MIC" just like in the original file. 

Original File: 
![image](https://github.com/mikeryan/crackle/assets/151716249/1f363866-9bd2-4a86-ba76-6d2f1765b5a5)

Decrypted File:
![image](https://github.com/mikeryan/crackle/assets/151716249/99e0af92-b381-4275-a05e-a29ed42f168e)

So my question is: Is the "decrypt with LTK" part of crackle able to decript LE Secure Connection when the LTK is known?   

Many Thanks!

**Comment by jkrauss1987** [comment]
I just saw that the decryption seems to work well but the problem is rather that the BT LE Link Layer and the BTL2CAP Layer are not recognized/parsed by Wireshark. I think that might be because the BTLE.CRC value is not recalculated but instead set to 0x 00 00 00. Is there a way to say Wireshark to ignore that? 

In addition i saw that crackle does not change the Length Field of the payload (nordic_ble.plen). I am using a nordic nRF BLE Sniffer. The second and third byte of the frame is for the payload length. Could this be implemented?

![image](https://github.com/mikeryan/crackle/assets/151716249/424414f9-2797-4e18-9d5a-e1eb0f560074)


Regards

**Comment by trentp-igor** [comment]
Decrypting LE Secure Connection with LTK doesn't seem to work at all for me.

Without supplying LTK, it fails as expected:
```
 Analyzing connection 0:
  44:01:bb:a0:d6:17 (public) -> 48:27:e2:2d:9a:66 (public)
  Found 23 encrypted packets
  Unable to crack due to the following error:
    LE Secure Connections
```

With LTK it just doesn't decrypt the packets for some reason:
```
[crackle]$ ./crackle -i ../pairing.pcapng -l AD1FE6CCEE0E20E701108DE0141BE9ED -o test.pcap
…
Analyzing connection 0:
  44:01:bb:a0:d6:17 (public) -> 48:27:e2:2d:9a:66 (public)
  Found 23 encrypted packets
  Decrypted 0 packets
```

**Comment by XenoKovah** [comment]
@trentp-igor FWIW it worked for me on a Just-Works Secure Connection when I put the endianness correct. I.e. currently Wireshark displays stuff little-endian, so if I just copy the LTK value from a Wireshark HCI log it doesn't work. But when I reversed the order of bytes to the big-endian the tool expects, it worked. Given that you're posting in caps and Wireshark doesn't use caps I'm guessing you didn't get the LTK from Wireshark, but still worth giving it a try to reverse the order.

**Comment by XenoKovah** [comment]
@jkrauss1987 given that you confirmed decryption works w/ LTK shouldn't you close this as actually working and just open a ticket for Wireshark people to ask how to ignore invalid CRC? (Which it sounds like was a problem from the source capture, not CrackLE. Perhaps use [Sniffle](https://github.com/nccgroup/Sniffle) instead of the nRF sniffer, since it calculates CRCs correctly?)

**Comment by jkrauss1987** [comment]
@XenoKovah you are right. The decription of crackle works with LTK, so i close the issue. Is far as there was no update, the BTLE.CRC value is not recalculated and the payload length is not updated after decryption what leads to a wrong parsing in Wireshark.
The bsniffhub tool does this correct but converting a pcap file here takes much time as it is not the main scope of the software...
Thanks for your help!

**Comment by trentp-igor** [comment]
It's been long enough I don't remember anymore, but I think I was getting the LTK from micropython.  The device was running micropython with the NimBLE stack.  The LTK is part of the NimBLE stack, but it gets passed up to micropython since you need to write code to store it somewhere, reload it, write your "delete all pairings" feature, etc.

I think I also tried getting the LTK from the other side by pairing to a Linux central running bluez, which stores the LTKs in some files in /var somewhere.

I could never get Nordic's sniffer firmware to decrypt packets via LTK either, which it is supposed to be able to do.

The only thing that ended up working was to build the firmware for the device using the option for debug device keys.  This uses a well known set of keys for the device itself, which the Nordic sniffer firmware knows about.
